### PR TITLE
Add ovirt_clusters and ovirt_clusters_facts modules

### PR DESCRIPTION
**Pull request by machacekondra**
##### ISSUE TYPE

- New Module Pull Request
##### COMPONENT NAME

ovirt_clusters
ovirt_clusters_facts
##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.1.1.0
```
##### SUMMARY

_ovirt_clusters_ module is module for manage oVirt clusters
_ovirt_clusters_facts_ module is module for fetch facts about oVirt clusters

```

```

### CODE REVIEW

**Comment by machacekondra** [comment]
@mwperina @ryansb

**Comment by machacekondra** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @machacekondra for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by machacekondra** [comment]
shipit

**Comment by gregdek** [comment]
Thanks @machacekondra. Since you are one of the maintainers (@machacekondra) of this module and you commented with "shipit", we are marking this PR for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Review by mwperina — APPROVED** [review]
Looks good to me, thanks! :+1:

File: cloud/ovirt/ovirt_clusters_facts.py
Comment (mwperina) [inline cloud/ovirt/ovirt_clusters_facts.py]: I'd rephrase to: 'Gather facts about all clusters which names start with C<production>:'

**Review by mwperina — CHANGES_REQUESTED** [review]
Sorry, I've missed that during previous review

File: cloud/ovirt/ovirt_clusters_facts.py
Comment (mwperina) [inline cloud/ovirt/ovirt_clusters_facts.py]: 'C(production*)' -> 'C(production)'

**Review by mwperina — CHANGES_REQUESTED** [review]

**Review by mwperina — APPROVED** [review]
Looks good to me, thanks! :+1:

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [inline cloud/ovirt/ovirt_clusters.py:553]: catchall exceptions are normally frowned upon as the error message here might not reflect the action attempted, we prefer friendlier error messages on each possible failure action that help the user figure out what to fix in order for the task to work.

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [inline cloud/ovirt/ovirt_clusters.py]: Since 2.2 we want to avoid \* imports to make modules lintable/testable

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [inline cloud/ovirt/ovirt_clusters.py:67]: this and the following descriptions are not clear on what input is needed here

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [inline cloud/ovirt/ovirt_clusters.py:4]: non redhat contributors should not set red hat copyright

**Comment by gregdek** [comment]
@machacekondra A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/ovirt/ovirt_clusters.py
Comment (machacekondra) [reply cloud/ovirt/ovirt_clusters.py:4]: This is Red Hat contribution.

File: cloud/ovirt/ovirt_clusters.py
Comment (machacekondra) [reply cloud/ovirt/ovirt_clusters.py]: Even for: `from ansible.module_utils.basic import *` ?

File: cloud/ovirt/ovirt_clusters.py
Comment (machacekondra) [reply cloud/ovirt/ovirt_clusters.py:553]: oVirt Python SDK returns friendly error message with description what's not correct in the request.

File: cloud/ovirt/ovirt_clusters.py
Comment (machacekondra) [reply cloud/ovirt/ovirt_clusters.py]: I will add there: `# flake8: noqa`

**Comment by machacekondra** [comment]
This module now passed flake8.
Documentation of parameters was improved.

**Comment by machacekondra** [comment]
shipit

**Review by mwperina — APPROVED** [review]
:+1:

**Comment by gregdek** [comment]
Thanks @machacekondra. Since you are one of the maintainers (@machacekondra) of this module and you commented with "shipit", we are marking this PR for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [reply cloud/ovirt/ovirt_clusters.py:4]: he, never mind then ....

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [inline cloud/ovirt/ovirt_clusters.py]: should switch bre restricted to the choices outlined in description?

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [reply cloud/ovirt/ovirt_clusters.py:553]: ok, if the library has good messages we can overlook this

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [reply cloud/ovirt/ovirt_clusters.py:67]: much better now

File: cloud/ovirt/ovirt_clusters.py
Comment (bcoca) [inline cloud/ovirt/ovirt_clusters.py]: We are moving away from \* imports, also no need to have this at the end of the file anymore (it is a real import now).

Comment (machacekondra) [reply cloud/ovirt/ovirt_clusters.py]: done

File: cloud/ovirt/ovirt_clusters.py
Comment (machacekondra) [reply cloud/ovirt/ovirt_clusters.py]: done

File: cloud/ovirt/ovirt_clusters.py
Comment (ryansb) [inline cloud/ovirt/ovirt_clusters.py:25]: You'll need a conditional here so that your module can raise a nice error instead of an `ImportError` stack trace. See the AWS modules like [this one](https://github.com/ansible/ansible-modules-core/blob/devel/cloud/amazon/cloudformation.py#L189-L192) for an example. Later, they check that `HAS_BOTO3` is set, then bomb out if it's `False`.

Comment (machacekondra) [reply cloud/ovirt/ovirt_clusters.py:25]: We check the SDK and raise friendly error [here](https://github.com/ansible/ansible-modules-extras/pull/3138/files#diff-cb60a765df80bc6d3e43ca8b069a7e2fR193)
which has [following](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ovirt.py#L47) implementation.
We use it so we can manage the error message and version check at one place.

Comment (machacekondra) [reply cloud/ovirt/ovirt_clusters.py:25]: Is that OK?

**Comment by mykaul** [comment]
Where and how can we define the rest of the cluster properties? For example, support ballooning, random number generator, etc.?

**Comment by machacekondra** [comment]
We decided to implement the basic one at the begging and add more when users request, but no problem to add those, before merge.

**Comment by machacekondra** [comment]
I've added following new parameters:

 - ballooning
 - gluster
 - threads_as_cores
 - ksm_numa
 - ksm
 - ha_reservation
 - trusted_service
 - vm_reason
 - host_reason
 - memory_policy
 - rng_sources

**Comment by machacekondra** [comment]
So, now, I've implemented all cluster attributes.

**Comment by gregdek** [comment]
Thanks @machacekondra for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by machacekondra** [comment]
shipit

**Comment by gregdek** [comment]
Thanks @machacekondra. Since you are one of the maintainers (@machacekondra) of this module and you commented with "shipit", we are marking this PR for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Review by mwperina — APPROVED** [review]
:+1:

**Pull request merged**
